atoma-auth/src/google.rs

```diff
@@ -0,0 +1,134 @@
+use std::collections::{HashMap, HashSet};
+
+use base64::{prelude::BASE64_URL_SAFE_NO_PAD, Engine};
+use jsonwebtoken::{decode, decode_header, Algorithm, DecodingKey, Validation};
+use reqwest::Client;
+use rsa::{pkcs1::EncodeRsaPublicKey, BigUint, RsaPublicKey};
+use serde::{Deserialize, Serialize};
+use serde_json::Value;
+use thiserror::Error;
+
+pub const ISS: &str = "https://accounts.google.com";
+
+/// Claims struct for Google ID tokens
+#[derive(Debug, Serialize, Deserialize)]
+pub struct Claims {
+    iss: String,               // Issuer
+    pub sub: String,           // Subject (user ID)
+    aud: String,               // Audience
+    exp: usize,                // Expiration time (as a UNIX timestamp)
+    pub email: Option<String>, // User's email (optional)
+}
+
+#[derive(Error, Debug)]
+pub enum GoogleError {
+    #[error("Base64 decode error: {0}")]
+    Base64DecodeError(#[from] base64::DecodeError),
+    #[error("RSA public key error: {0}")]
+    RsaPublicKeyError(#[from] rsa::errors::Error),
+    #[error("Decoding key error: {0}")]
+    DecodingKeyError(#[from] jsonwebtoken::errors::Error),
+    #[error("Request error: {0}")]
+    RequestError(#[from] reqwest::Error),
+    #[error("Serde JSON error: {0}")]
+    SerdeJsonError(#[from] serde_json::Error),
+    #[error("PEM error: {0}")]
+    PemError(#[from] rsa::pkcs1::Error),
+    #[error("Missing kid in token header")]
+    MissingKid,
+    #[error("Public keys malformed")]
+    MalformedPublicKeys,
+    #[error("Email not found in token claims")]
+    EmailNotFound,
+}
+
+type Result<T> = std::result::Result<T, GoogleError>;
+
+/// Convert a JWK to a DecodingKey
+fn jwk_to_decoding_key(n: &str, e: &str) -> Result<DecodingKey> {
+    // Decode `n` and `e` from Base64URL
+    let n_bytes = BASE64_URL_SAFE_NO_PAD.decode(n)?;
+    let e_bytes = BASE64_URL_SAFE_NO_PAD.decode(e)?;
+
+    // Convert `n` and `e` into BigUint
+    let n = BigUint::from_bytes_be(&n_bytes);
+    let e = BigUint::from_bytes_be(&e_bytes);
+
+    // Construct the RSA public key
+    let public_key = RsaPublicKey::new(n, e)?;
+
+    // Convert the RSA key to PEM format
+    let pem = public_key.to_pkcs1_pem(rsa::pkcs1::LineEnding::LF)?;
+
+    // Create the DecodingKey
+    let decoding_key = DecodingKey::from_rsa_pem(pem.as_bytes())?;
+
+    Ok(decoding_key)
+}
+
+/// Fetch Google's public keys from the JWKS endpoint
+pub async fn fetch_google_public_keys() -> Result<HashMap<String, DecodingKey>> {
+    let jwks_url = "https://www.googleapis.com/oauth2/v3/certs";
+    let client = Client::new();
+    let res = client.get(jwks_url).send().await?;
+    let jwks: Value = res.json().await?;
+    let keys = jwks["keys"]
```

Comment:
Add a constant for this `KEYS`. Also, I suggest using `.get(KEYS)` to avoid panics on the proxy